### ISSUE TITLE
Micro::Cases.flow(transaction: true)

### ISSUE THREAD

**Issue by serradura** [serradura/u-case#44]
Wraps a `Micro::Cases::Flow` inside of an` ActiveRecord::Transactions` and if an exception happening or any step returns a failure, this flow will be halted because an `ActiveRecord::Rollback` will be raised.

```ruby
MyFlow = Micro::Cases.flow([
  NormalizeParams,
  ValidatePassword,
  Micro::Cases.flow(db_transaction: true, steps: [
    CreateUser,
    CreateUserProfile
  ]),
  EnqueueIndexingJob
])

# ---

MyFlowWrappedInATransaction = Micro::Cases.flow(db_transaction: true, steps: [
  CreateUser,
  CeateUserProfile
])

MyFlow = Micro::Cases.flow([
  NormalizeParams,
  ValidatePassword,
  MyFlowWrappedInATransaction,
  EnqueueIndexingJob
])
```

```ruby
class MyFlow < Micro::Case
  flow(db_transaction: true, steps: [
    NormalizeParams,
    ValidatePassword,
    CreateUser,
    CreateUserProfile
  ])
])

# ---

MyFlowWrappedInATransaction =
  Micro::Cases.flow(db_transaction: true, steps: [
    CreateUser,
    CeateUserProfile
  ])

class MyFlow < Micro::Case
  flow([
    NormalizeParams,
    ValidatePassword,
    MyFlowWrappedInATransaction
  ])
])
```

# Definition of done:
- [ ] This mode/plugin will be disabled by default, that is, it will be enabled to be available. Except for its core dependencies `kind` and `u-attributes`, this gem never will require any external dependency by default.

---

Thanks, @marlosirapuan, @josuetex, @marcosgz, @MatheusRich, @mrbongiolo for helping me to elaborate on this idea. 🚀

**Comment by marlosirapuan** [comment]
why not just `transaction`? cause it's already a flow.. 🤔

**Comment by serradura** [comment]
@marlosirapuan I'm not sure about the method name yet. But this mode could be enabled using a kwarg. e.g:

```ruby
Micro::Cases.flow(transaction: true, [
  CreateUser,
  CeateUserProfile
])

# --

class MyCase < Micro::Case
  flow(transaction: true, [
  CreateUser,
  CeateUserProfile
])
```

What do you think? cc: @MatheusRich

**Comment by MatheusRich** [comment]
To me `flow_in_a_db_transaction` is too much verbose.

An alternative:

```ruby
MyFlowWrappedInATransaction = Micro::Cases.flow(db_transaction: true, [
  CreateUser,
  CeateUserProfile
])
```

**Comment by serradura** [comment]
@MatheusRich I liked the usage of `db_transaction` kwarg. Nice!

**Comment by mrbongiolo** [comment]
@serradura how about?
```ruby
MyFlow = Micro::Cases.flow(transaction: :db, [])
# or
MyFlow = Micro::Cases.flow(transaction: Micro::Cases::DB_TRANSACTION, [])
# or
MyFlow = Micro::Cases.flow(transaction: MyCustomImplementation, [])
```

**Comment by serradura** [comment]
@mrbongiolo I can think of different types of transactions.  But I would like to start solving this first one: DB transactions. As you proposed, I believe that if a new kind of transaction appears, we could make `db_transation: true` become an alias for `transaction: :db`. But its an excellent suggestion. Thank you!

**Comment by marcosgz** [comment]
@serradura I totally agree with @mrbongiolo suggestion of `:transaction` naming: `MyFlow = Micro::Cases.flow(transaction: MyCustomImplementation, [])`

I've worked on applications that had more than one (ActiveRecord)[https://github.com/rails/rails/tree/master/activerecord] or (Sequel)[https://github.com/jeremyevans/sequel) db connections. The transaction itself would not even need to be part of the project. And just provide mechanisms to group the a list of cases in a single transaction. And good examples in the documentation.

I don't think someone is using two ORM like this in same application. But a generic API would make that possible:

```ruby
class SequelTransaction < Micro::Cases::Transaction
  def call!
    result = nil
    SequelDB.transaction do
      result = yield
      raise Sequel::Rollback if result.failure?
    end
    result
  end
end

class ActiveRecordTransaction < Micro::Cases::Transaction
  def call!
    result = nil
    ApplicationRecord.transaction do
      result = yield
      raise ActiveRecord::Rollback if result.failure?
    end
    result
  end
end

SignInFlow = Micro::Cases.flow(transaction: ActiveRecordTransaction, [CreateUser, CreateProfile])
SubscribeFlow = Micro::Cases.flow(transaction: SequelTransaction, [CreateOrder, CreateSubscription])
```
Ignore the syntax itself. Just the idea of a generic transaction.

**Comment by mrbongiolo** [comment]
I was thinking exactly in that way, so the gem could have its own AR db transaction implementation (just for easy of use), but it could receive any object that respects the u-case transaction API.
Github <notifications@github.com> wrote:
“@serradura I totally agree with @mrbongiolo suggestion of :transaction naming: MyFlow = Micro::Cases.flow(transaction: MyCustomImplementation, [])

I've worked on applications that had more than one (ActiveRecord)[https://github.com/rails/rails/tree/master/activerecord] or (Sequel)[https://github.com/jeremyevans/sequel) db connections. The transaction itself would not even need to be part of the project. And just provide mechanisms to group the a list of cases in a single transaction. And good examples in the documentation.

I don't think someone is using two ORM like this in same application. But a generic API would make that possible:

class SequelTransaction < Micro::Cases::Transaction
  def call!
    result = nil
    SequelDB.transaction do
      result = yield
      raise Sequel::Rollback if result.failure?
    end
    result
  end
end

class ActiveRecordTransaction < Micro::Cases::Transaction
  def call!
    result = nil
    ApplicationRecord.transaction do
      result = yield
      raise ActiveRecord::Rollback if result.failure?
    end
    result
  end
end

SignInFlow = Micro::Cases.flow(transaction: ActiveRecordTransaction, [CreateUser, CreateProfile])
SubscribeFlow = Micro::Cases.flow(transaction: SequelTransaction, [CreateOrder, CreateSubscription])Ignore the syntax itself. Just the idea of a generic transaction.”

**Comment by PavarinE** [comment]
Hello @serradura . Do we have any plans to add this feature still? I was looking exactly for support to db transactions in a flow.

**Comment by wbotelhos** [comment]
@serradura Any plans to release this feature?

**Comment by serradura** [comment]
Hey @marlosirapuan, @josuetex, @MatheusRich, @marcosgz, @mrbongiolo, @PavarinE, @wbotelhos 👋 — finally landed in #143.

Final shape:

```ruby
# enable with the default class
Micro::Cases.flow(transaction: true, steps: [...])

# multi-DB — explicit per call
Micro::Cases.flow(transaction: { with: AnalyticsRecord }, steps: [...])

# class-level declaration (inherited like `flow` / `attributes`)
class SignUp < Micro::Case
  transaction with: ApplicationRecord
  flow(transaction: true, steps: [...])
end

# inline inside call!
transaction(with: ApplicationRecord) { call(CreateUser).then(CreateProfile) }

# global default — set once in a Rails initializer
Micro::Case.config do |config|
  config.default_transaction_class { ApplicationRecord }
end
```

A couple of decisions worth flagging back to the thread:

- @marlosirapuan's instinct on `transaction` (vs `db_transaction`) won out — it's the flow's own concept, not a separate plugin.
- The `with:` vocabulary picks up the multi-DB story that @mrbongiolo and @marcosgz raised. For now it requires a subclass of `ActiveRecord::Base` because the rollback path hardcodes `ActiveRecord::Rollback`, but the `{ with: ... }` shape leaves the door open to plug in other adapters (Sequel, custom transaction objects) later without breaking the public API.
- The "no AR dependency by default" definition-of-done from the issue is preserved — the gem still doesn't `require 'active_record'`. Calling `transaction: true` without AR loaded raises a curated `Micro::Cases::Error::TransactionAdapterMissing` so the misconfig surfaces immediately instead of cratering on a NameError later.

Thanks for all the back-and-forth — the API is genuinely better for it 🙏
